client-api/src/main/java/org/ostelco/topup/api/db/SubscriberDAOInMemoryImpl.java

```diff
@@ -0,0 +1,85 @@
+package org.ostelco.topup.api.db;
+
+import org.ostelco.topup.api.core.Consent;
+import org.ostelco.topup.api.core.Error;
+import org.ostelco.topup.api.core.Product;
+import org.ostelco.topup.api.core.Profile;
+import org.ostelco.topup.api.core.SubscriptionStatus;
+
+import io.vavr.control.Either;
+import io.vavr.control.Option;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import lombok.NonNull;
+
+/**
+ * An 'in-memory' store for testing.
+ *
+ */
+public class SubscriberDAOInMemoryImpl implements SubscriberDAO {
+
+    /* Table for 'profiles'. */
+    private final Map<String, Profile> profileTable = new HashMap<>();
```

Comment:
Actually, this needs to be thread safe.  The requests coming into the storage may produced by multiple threads, and may overlap in time.  I suggest either using a concurrent datastruture (e.g. https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html implemented in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html), or to put a lock around  the accesses to the profileTable.   A simple "spin lock" (e.g. an Object instance used in a "synchronised" block) can easily lead to lock contention, but is is simple to implement, so if you can't easily fit in a concurrent datastructure I suggest using a spin lock.   For production use, if  a concurrent datastructure isn't usable, I would recommend using read/write locks, that can be quite efficient if done right.    The in-memory thing won't be used in production so that isn't very important, but avoiding race conditions even in test is .-)